### PR TITLE
Added end-to-end MNIST tensor training test.

### PR DESCRIPTION
**Pull request by dlibenzi**

### CODE REVIEW

**Comment by dlibenzi** [comment]
Hmm, loss goes down, accuracy does not pick up.
Lemme check ...

**Comment by dlibenzi** [comment]
OK, there is an issue with the ops we use for accuracy testing.
If we move result to CPU and do the ops there, accuracy is fine.
```
      data = data.to(device=device)
      # target = target.to(device=device)                                                                                                                                                                                                   

      y = model(data).cpu()
      test_loss += loss_fn(y, target).sum().item()
      pred = y.max(1, keepdim=True)[1]
      correct_count += pred.eq(target.view_as(pred)).sum().item()
      count += FLAGS.batch_size
```

File: test/test_train_mnist_tensor.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+      if data.size()[0] != FLAGS.batch_size:
+        break
+      data = data.to(device=device)
+      target = target.to(device=device)
+
+      optimizer.zero_grad()
+      y = model(data)
+      loss = loss_fn(y, target)
+      loss.backward()
+      xm.optimizer_step(optimizer)
```
Comment (asuhan) [inline test/test_train_mnist_tensor.py]: What do you think about making this look more idiomatic? Instead of calling the `optimizer_step` like this, have a wrapper class with a step method. No need to address this right now, just curious whether there's a fundamental reason you went for this instead.

File: torch_xla_py/xla_model.py [hunk 2]
```diff
@@ -355,6 +360,29 @@ def xla_run_grad(xla_model, grad_outputs, devices=None):
   xla_model.backward(*grads_output_xla)
 
 
+def sync_optimizer(optimizer):
```
Comment (asuhan) [inline torch_xla_py/xla_model.py]: This is hard-wired to dispatch the accumulated graph at the end of an optimizer step. Could we do something more general instead which would also work for a hook on forward? It should be possible to do this with no knowledge of the user code except for the fact that a training loop is complete. Again, not something for this PR, just a question for my own understanding.

Comment (dlibenzi) [reply torch_xla_py/xla_model.py]: Sure, that is another way. Though, if the user wants to interact with the optimizer core object (calling other APIs, accessing its fields), his code needs to be changed in many places.

File: test/test_train_mnist_tensor.py (hunk 1 above)
Comment (dlibenzi) [reply test/test_train_mnist_tensor.py]: Sure, that is another way. Though, if the user wants to interact with the optimizer core object (calling other APIs, accessing its fields), his code needs to be changed in many places.

(comment in wrong place)

File: torch_xla_py/xla_model.py (hunk 2 above)
Comment (dlibenzi) [reply torch_xla_py/xla_model.py]: Hooking the model is another place. Again, this means that is the user wants to interact with the model's other APIs/fields, he needs to change more.
With the optimizer_step(), only only line is changed.
But there are many ways we can go.
Bottom line, we need to "sync" parameters and gradients, in order to make a repeatable computation.

**Comment by dlibenzi** [comment]
Yes, working on the accuracy issues ATM.
Hold on...

File: torch_xla_py/xla_model.py (hunk 2 above)
Comment (dlibenzi) [reply torch_xla_py/xla_model.py]: Correction, model wrapping is not good (unless we pass optimizer - which is a chicken and egg problem), as we need to sync the optimizer's state as well (think, momentum).

Comment (asuhan) [reply torch_xla_py/xla_model.py]: Yeah, these are good points. As long as there's no fundamental reason why this couldn't be more general, I'm not worried. We can always rework the Python APIs if we change our mind.

**Comment by dlibenzi** [comment]
Never mind, looks like an artifact of learning rate and batch size.
I think we can go in now, but before enabling the test we need to double check.
